internal/lang/funcs/cidr_test.go

```diff
@@ -396,3 +396,107 @@ func TestCidrSubnets(t *testing.T) {
 		})
 	}
 }
+
+func TestCidrContains(t *testing.T) {
+	noError := func(err error) bool { return err == nil }
+
+	tests := []struct {
+		Prefix  cty.Value
+		Address cty.Value
+		Want    cty.Value
+		ErrFn   func(error) bool
+	}{
+		{
+			// IPv4, contained (IP).
+			cty.StringVal("192.168.2.0/20"),
+			cty.StringVal("192.168.2.1"),
+			cty.True,
+			noError,
+		},
+		{
+			// IPv4, contained (CIDR).
+			cty.StringVal("192.168.2.0/20"),
+			cty.StringVal("192.168.2.0/22"),
+			cty.True,
+			noError,
+		},
+		{
+			// IPv4, not contained.
+			cty.StringVal("192.168.2.0/20"),
+			cty.StringVal("192.126.2.1"),
+			cty.False,
+			noError,
+		},
+		{
+			// IPv4, not contained (CIDR).
+			cty.StringVal("192.168.2.0/20"),
+			cty.StringVal("192.126.2.0/18"),
+			cty.False,
+			noError,
+		},
+		{
+			// IPv6, contained.
+			cty.StringVal("fe80::/48"),
+			cty.StringVal("fe80::1"),
+			cty.True,
+			noError,
+		},
+		{
+			// IPv6, not contained.
+			cty.StringVal("fe80::/48"),
+			cty.StringVal("fe81::1"),
+			cty.False,
+			noError,
+		},
+		{
+			// Address family mismatch: IPv4 prefix, IPv6 address.
+			cty.StringVal("192.168.2.0/20"),
+			cty.StringVal("fe80::1"),
+			cty.NilVal,
+			func(err error) bool {
+				return err != nil && err.Error() == "address family mismatch: 192.168.2.0/20 vs. fe80::1"
+			},
+		},
+		{
+			// Address family mismatch: IPv6 prefix, IPv4 address.
+			cty.StringVal("fe80::/48"),
+			cty.StringVal("192.168.2.1"),
+			cty.NilVal,
+			func(err error) bool {
+				return err != nil && err.Error() == "address family mismatch: fe80::/48 vs. 192.168.2.1"
+			},
+		},
+		{
+			// Input error: invalid CIDR address.
+			cty.StringVal("not-a-cidr"),
+			cty.StringVal("192.168.2.1"),
+			cty.NilVal,
+			func(err error) bool {
+				return err != nil && err.Error() == "invalid CIDR address: not-a-cidr"
+			},
+		},
+		{
+			// Input error: invalid IP address.
+			cty.StringVal("192.168.2.0/20"),
+			cty.StringVal("not-an-address"),
+			cty.NilVal,
+			func(err error) bool {
+				return err != nil && err.Error() == "invalid IP address or prefix: not-an-address"
+			},
+		},
+	}
```

Comment:
Interesting question. Should the function actually check for such misrepresentations?

I'm personally leaning towards a "no" because it feels like a separate functionality - perhaps one that warrants its own function (that could be used for variable validations for example). What's your take?

Reply:
Testing is all about paranoia, so let's add the test.  In providing the test, we're implicitly promising future users that we'll handle this case in this specific manner.  Those two reasons seem justification enough.

That said, I'd comment the test along the lines of:  we rely on ipaddr for thie behaviour.  If the test starts failing, it's likely because of a downstream change.  ipaddr already [checks for this situation](https://github.com/opentffoundation/opentf/blob/6d9c42cd2f345b0c29f5417daeea727230d0221e/internal/ipaddr/ip_test.go#L80).  As the test already exists in ipaddr, that package is implicitly promising to handle it for us.  If we started finding errors around this, then we'd likely go inspect the package and fix that.

Reply:
OK sorry I misunderstood. Sure, I'm all for adding a test.
I thought you suggested that the function errors on a misrepresented prefix.